### PR TITLE
[naga msl-out] Avoid UB by making all loops bounded.

### LINKED ISSUE #6528: [naga msl-out] Unbounded loop workaround is not adequate
I'll provide more details later, but the `LOOP_IS_REACHABLE` kludge has been found to be insufficient at avoiding UB and security problems. Basically, even with the kludge applied, the loop is still infinite, and still reached, so it's just a matter of being clever.

- [ ] msl
- [ ] hlsl

### PR DESCRIPTION
**Pull request by jimblandy**
In MSL output, avoid undefined behavior due to unbounded loops by adding an unpredictable, never-actually-taken `break` to the bottom of each loop body, rather than adding an unpredictable, never-actually-taken branch over each loop.

This will probably have more of a performance impact, because it affects each iteration of the loop, but unlike branching over the loop, which leaves infinite loops (and thus undefined behavior) in the output, this actually ensures that no loop presented to Metal is unbounded, so that there is no undefined behavior present that the optimizer could use to make unwelcome inferences.

Fixes #6528.

### CODE REVIEW

**Comment by JMS55** [comment]
Is this something we can turn off via wgpu/naga for native platforms?

**Comment by cwfitzgerald** [comment]
Yes, see https://github.com/gfx-rs/wgpu/pull/6520

**Review by teoxoy — APPROVED** [review]
LGTM

File: naga/src/back/msl/writer.rs [hunk 1]
```diff
@@ -764,52 +765,51 @@ impl<W: Write> Writer<W> {
...
-    /// would be to add inline assembly to each loop that is annotated as
-    /// potentially branching out of the loop, but which in fact generates no
-    /// instructions. Unfortunately, inline assembly is not handled correctly by
-    /// some Metal device drivers. Further experimentation hasn't produced a
-    /// satisfactory approach.
+    /// To avoid UB, Naga must persuade the MSL compiler that no loop Naga
+    /// generates is infinite. One approach would be to add inline assembly to
+    /// each loop that is annotated as potentially branching out of the loop,
+    /// but which in fact generates no instructions. Unfortunately, inline
+    /// assembly is not handled correctly by some Metal device drivers.
```
Comment (DemiMarie) [inline naga/src/back/msl/writer.rs:772]: Which drivers are these?

File: naga/src/back/msl/writer.rs [hunk 2]
```diff
@@ -764,52 +765,51 @@ impl<W: Write> Writer<W> {
...
-        let loop_reachable_volatile_name = self.namer.call("unpredictable_jump_over_loop");
+        self.force_bounded_loop_macro_name = self.namer.call("LOOP_IS_BOUNDED");
+        let loop_bounded_volatile_name = self.namer.call("unpredictable_break_from_loop");
         writeln!(
             self.out,
-            "#define {} if (volatile bool {} = true; {})",
-            self.loop_reachable_macro_name,
-            loop_reachable_volatile_name,
-            loop_reachable_volatile_name,
+            "#define {} {{ volatile bool {} = false; if ({}) break; }}",
```
Comment (DemiMarie) [inline naga/src/back/msl/writer.rs:809]:
```suggestion
            "#define {} if (true) {{ volatile bool {} = false; if ({}) break; }} else do {{}} while (0)",
```

This is what I would use in hand-written code, as it forces the macro to be followed by a semicolon, but in generated code I imagine that simplicity is more important.

File: naga/src/back/msl/writer.rs (hunk 1 above)
Comment (jimblandy) [reply naga/src/back/msl/writer.rs:772]: I never learned exactly which drivers.

File: naga/src/back/msl/writer.rs (hunk 2 above)
Comment (jimblandy) [reply naga/src/back/msl/writer.rs:809]: Yes - we control the vertical and the horizontal here, so legibility for the unfortunate souls who have to read our output comes first.

**Pull request merged**